bliss/hyperparameter.py

```diff
@@ -0,0 +1,315 @@
+            else self.enc_conv_c
+        )
+
+        self.cfg.model.encoder.params["enc_kern"] = (
+            trial.suggest_int(
+                "enc_kern",
+                self.enc_kern[0],
+                self.enc_kern[1],
+                self.enc_kern[2],
+            )
+            if type(self.enc_kern) is not int
+            else self.enc_kern
+        )
+
+        self.cfg.model.encoder.params["enc_hidden"] = (
+            trial.suggest_int(
+                "enc_hidden",
+                self.enc_hidden[0],
+                self.enc_hidden[1],
+                self.enc_hidden[2],
+            )
+            if type(self.enc_hidden) is not int
+            else self.enc_hidden
+        )
+
+        lr = (
+            trial.suggest_loguniform("learning rate", self.lr[0], self.lr[1])
+            if type(self.lr) is not float
+            else self.lr
+        )
+        weight_decay = (
+            trial.suggest_loguniform(
+                "weight_decay", self.weight_decay[0], self.weight_decay[1]
+            )
+            if type(self.weight_decay) is not float
+            else self.weight_decay
+        )
+
+        # set up the optimizer base on learning rate and weight decay
+        self.cfg.optimizer.params.update(dict(lr=lr, weight_decay=weight_decay))
+
+        # set up the checkpoint callback for the model
+        checkpoint_callback = pl.callbacks.ModelCheckpoint(
+            os.path.join(self.model_dir, "trial_{}".format(trial.number), "{epoch}"),
+            monitor=self.monitor,
+        )
+
+        # Initiate the model
+        model = SleepPhase(self.cfg)
+
+        # put correct device to model
+        use_gpu = 0
+        if self.gpu_queue is not None and self.single_gpu is None:
+            use_gpu = [gpu_id]
+        elif self.gpu_queue is None and self.single_gpu is not None:
+            use_gpu = [self.single_gpu]
+
+        # put gpu id back to the queue
+        if self.gpu_queue is not None:
+            self.gpu_queue.put(gpu_id)
+
+        # set up the trainer
+        ## PyTorchLightningPruningCallback allow the pruner to stop early
+        trainer = pl.Trainer(
+            logger=False,
+            gpus=use_gpu,
+            checkpoint_callback=checkpoint_callback,
+            max_epochs=self.max_epochs,
+            callbacks=[
+                self.metrics_callback,
+                PyTorchLightningPruningCallback(trial, monitor=self.monitor),
+            ],
+        )
+
+        # start training
+        trainer.fit(model, datamodule=star_dataset)
+
+        return self.metrics_callback.metrics[-1][self.monitor].item()
+
+
+def SleepTune(
+    cfg: DictConfig,
+    max_epochs: int,
+    model_dir,
+    monitor,
+    n_batches,
+    batch_size,
+    direction: str = "minimize",
+    data_seed: int = 10,
+    n_trials=100,
+    time_out=600,
+    gc_after_trial: bool = False,
+    sampler=None,
+    pruner=None,
+    num_gpu: int or list = 2,
+    storage=None,
```

Comment:
since this class has many keyword arguments, it would be great if you created a config file for optuna/hyper-parameter tuning that contains these parameters. I'm also ok doing this in a future PR but I just thought I mentioned it as an idea.